### PR TITLE
Add `read_only` option to `FieldPanel`

### PR DESCRIPTION
**Pull request by ababic**
Related issue/discussion: #2893

Supersedes: #6194 and #6067

### When would you use this?

When you want to show a model field value in a form without it ever being editable, use `read_only=True` in the `FieldPanel` constructor, e.g. `FieldPanel("slug", read_only=True)`. 

### What is displayed in read-only mode?

Following the "Form follows function" design principle, if we don't intend for the value to be updated, there's no point in presenting the user with a form control (whether hidden, disabled or in any other special state). So, for read-only panels, we avoid adding fields to the (dynamically generated) `ModelForm`, and simplify the output.

Instead of a form control, we render the field value in a `<div>` element, which looks a lot like a disabled `<input type="text">` element, but is screen-reader navigable in much the same way as a regular (non-disabled) input.

While setting the `disabled` attribute on the Django form field would grant us similar protection, not including a field in the form at all gives us a cast-iron guarantee that field values included in POSTed data will not be applied to the model (unexpected values are simply ignored and will never be included in `form.cleaned_data`). Validation isn't applied to non-fields values either.

### Can I use it to display arbitrary attribute values?

I'm afraid not. `FieldPanel` has a pre-existing restriction that allows it to be used for model fields only (which makes sense when you consider the name). For arbitrary model instance values, it makes sense to introduce an entirely new panel type, which will contributed via a separate PR.

On the plus side: Having a model field allows us to pull `verbose_name` and `help_text` values from there and display them in a familiar manner.

### Are all field types supported?

Yes, but displaying chunks of user-controlled HTML mid-form is a touch on the risky side so, for `StreamField` or `RichTextField` values, a text-only representation is extracted from the HTML (using the same method as used for the 'compare revisions' view) and used for display instead.

If for any reason you don't like the default representation of a field value, you can override the `format_value_for_display()` method to control what is rendered.

```
class KeyIngredientsPanel(FieldPanel):
    """
    A custom FieldPanel for use with the 'key_ingredients' ParentalManyToManyField 
    """

    def format_value_for_display(self, value):
        # Show a comma-separate list of ingredients when in read-only mode
        return ", ".join(obj.get_listing_summary() for obj in value.all())
```

### How it looks

As children of a `MultiFieldPanel`:

Screenshot 2022-09-07 at 1 06 47 pm

As top-level `FieldPanel` instances:

Screenshot 2022-09-07 at 1 07 20 pm

### CODE REVIEW

File: wagtail/admin/panels.py
Comment (gasman) [inline wagtail/admin/panels.py:759]: Why would the field be present? Doesn't returning `{}` from `get_form_options` account for that (unless it's being inserted by some other panel, in which case we should probably trust that the other panel knows what it's doing)?

**Review by gasman — CHANGES_REQUESTED** [review]

File: wagtail/admin/panels.py
Comment (ababic) [reply wagtail/admin/panels.py:341]: Previously, I have found that the `instance._state.adding` check doesn't work for pages. I think because the instance are generally assembled in-memory from revision content. That may have been fixed since, though.

File: wagtail/admin/panels.py
Comment (ababic) [reply wagtail/admin/panels.py:759]: When `read_only_on_update` is used, we don't know whether the field is needed until we can tell if the instance is new or not. `get_form_options()` doesn't have that awareness, so we can't do that check there.

Comment (ababic) [reply wagtail/admin/panels.py:759]: I was thinking that `FieldPanel.BoundPanel.read_only` might possibly be updated to cater for other cases in future too, such as rendering a read-only value when a `permission` check fails - in which case, you wouldn't be able to figure that out until this point either.

Comment (gasman) [reply wagtail/admin/panels.py:759]: Thanks for the clarification! I'll need to spend some time thinking about the best way forward here - the whole premise of the panels refactor in 2.15 was that the Panel object specifies the makeup of the form class it wants to work with, and the BoundPanel then renders the resulting form. If BoundPanel is tinkering with the form fields after the fact, then that design has failed in its aim.

I'm also very reluctant to introduce a new 'hidden' contract between the view and the panels framework about the order things have to happen in. (In fact, if I'm not mistaken - for a successful POST action, we never call get_bound_panel at all, since no template is being rendered in that case, right? Which means it's unavoidable that form validation happens without having first triggered this logic.)

For the `FieldPanel(permission='...')` implementation, I took the stance that if the set of fields is going to vary dynamically, then that custom logic belongs inside the form class rather than in panels, hence the creation of [django-permissionedforms](https://github.com/wagtail/django-permissionedforms). But then that has its own issues - firstly, in this case, we want the form class to have as little involvement as possible with the read-only fields. And secondly, if the form class has to accommodate every conceivable variable that might alter the form (namely, the user in the case of django-permissionedforms, and the create-or-edit flag here), we'll end up with an unmaintainable tower of mixins.

In this case, I think the neatest approach might be to support the create and edit cases using different form classes - i.e. `Panel.get_form_class` and `Panel.get_form_options` will receive the create-or-edit flag and construct a form class tailored to the create or edit case respectively. (Right now we do create a fresh form class on every request, but deliberately limit the information that that process has access to, to avoid the pre-2.15 antipattern where the form class was being assembled on-the-fly and tailored to that specific request, user and object instance. In the current setup, where get_form_options takes no parameters, we're guaranteed that the form class will always be the same for a given model. At some point I'd like to cache those form classes across requests, but haven't got round to it yet... effectively, this change would be relaxing the "form class will always be the same" condition in a very limited and controlled way that leaves open the possibility of caching.)

Comment (ababic) [reply wagtail/admin/panels.py:759]: Nicely explained, thank you @gasman. Okay, I will take a look at implementing a `for_create` flag (open to suggestions on naming) for `Panel.get_form_class` and  `Panel.get_form_options`.

File: wagtail/admin/panels.py
Comment (ababic) [reply wagtail/admin/panels.py]: Since we are moving away from 'detection' in this way, I am marking this as resolved.

**Comment by ababic** [comment]
@gasman Have updated, revised the tests, and rebased for you :)

File: wagtail/admin/panels.py
Comment (ababic) [reply wagtail/admin/panels.py]: Resolved in 73a8aa99932588fa568be6ca91d8616d798f6e83

**Comment by ababic** [comment]
@gasman Was wondering what you think about handling `RichTextField` and `StreamField` values here? I would guess a clump of (escaped or not) HTML wouldn't be a great out-of-the-box experience here. Is it enough to document these limitations, or should we think about doing something different for those field types by default (e.g. strip out the text and only show that)?

**Comment by ababic** [comment]
Have gone with updating `FieldPanel.BoundPanel.is_shown()` to return `False` when a read-only view of a `RichTextField` or `StreamField` is requested.

**Comment by ababic** [comment]
@gasman I believe I've addressed all feedback here. Is there anything else preventing this from being merged?

**Comment by lb-** [comment]
Flagged for another review, not sure on timing @ababic sorry

**Comment by ababic** [comment]
@lb- much appreciated, thank you!

File: wagtail/admin/panels.py
Comment (ababic) [reply wagtail/admin/panels.py:318]: Thanks @thibaudcolas. The diff representation for richtext might be an option. I'll look into that! Following discussion with @gasman, I've steared clear of using form fields at all for read-only panels, so the 'read-only editor' approach would complicate things somewhat.

File: wagtail/admin/templates/wagtailadmin/panels/read_only_field_panel.html
Comment (ababic) [reply wagtail/admin/templates/wagtailadmin/panels/read_only_field_panel.html:10]: Yes, I figured showing a comment option next to something that is unlikely to ever change between revisions might be misleading from a review perspective. It's certainly not something I have strongly held views on, though.

File: wagtail/admin/panels.py
Comment (ababic) [reply wagtail/admin/panels.py:846]: Thanks @thibaudcolas, I'll explore this some more. From what I remember, using `id_for_label` resulted in some kind of unwanted output in one of the templates, but I might be able to get around that by changing what is included in the context.

File: wagtail/admin/templates/wagtailadmin/panels/read_only_field_panel.html
Comment (brylie) [reply wagtail/admin/templates/wagtailadmin/panels/read_only_field_panel.html:10]: The help text might be used to provide descriptive information about the read-only value. E.g., "use this order number to cross-reference the PayPal transaction" in an eCommerce application.

The `FieldPanel` should have consistent behavior even when adding different modalities. Since  developers are familiar with `help_text`, there is a somewhat high likelihood  that a use case for a `read_only` field will intersect with one for using `help_text`.

Just my two cents 🙂

File: wagtail/admin/panels.py
Comment (ababic) [reply wagtail/admin/panels.py]: > From what I remember, using `id_for_label` resulted in some kind of unwanted output in one of the templates

It was `multi_field_panel_child.html` I was referring to. I ditched the returning so that read-only children had a `<h3>` instead of a `<label>` element. Have since updated the conditional to cater for the read-only case.

> I think this is off (unsure why). In my demo site, I get a different id and aria-labelledby

This should be sorted now for both 'top-level' panels and children of `MultiFieldPanel` :)

File: wagtail/admin/templates/wagtailadmin/panels/read_only_field_panel.html
Comment (ababic) [reply wagtail/admin/templates/wagtailadmin/panels/read_only_field_panel.html]: Thanks for the input both! Have left comments showing by default, and devs can use the `disable_comments` option like normal to disable them.

**Comment by ababic** [comment]
@thibaudcolas I believe I've addressed all of your feedback now, so this is ready to look at once more.

**Comment by cam-yates** [comment]
Hey @ababic 

We've been using this code temporarily whilst waiting for it to be merged into a full release

One thing that we picked up on was fields which have had a `choice` selected still display the value and not the label which for us was less intuitive to the user

as a fix to that i've added the following to the start of `format_value_for_display()`
```
if self.db_field.choices:
    return dict(self.db_field.choices)[value]
```

**Comment by ababic** [comment]
@cam-yates that's really useful feedback. Thanks for trying this out!

**Comment by ababic** [comment]
Hi folks. Although github says there are no conflicts here, `main` has since undergone a major restructuring of panel code, so am in the process of reworking this to fit. When that is done, I would be hugely appreciative if someone from the core team could work with me to push this through? It's incredibly difficult not to take things personally when every contribution you make languishes in review hell for months.

**Comment by tomdyson** [comment]
@ababic I'm not qualified to help push it through but just want to pass on my huge appreciation for this work! I've added it to the agenda for the next core team meeting.

**Comment by ababic** [comment]
Thanks @tomdyson! 

I'm actually struggling to figure out how I had `read_only_for_update` working properly before, as there doesn't appear to be anywhere where I have access to the panel instance and the `is_editing` boolean from the view where I can safely impact the result without changing the panel instance's state (which is obviously undesirable).

I feel like the best thing for now is to remove `read_only_on_update` and the related changes for now, and figure it out at a later time.

File: wagtail/admin/panels/field_panel.py
Comment (gasman) [inline wagtail/admin/panels/field_panel.py:202]: Have these changes been introduced by mistake in a rebase? They seem to be unrelated to the rest of the PR, and partially match the changes made in #10320.

**Review by gasman — COMMENTED** [review]
Had some trouble rebasing this on current main, but I think I've got a working version at https://github.com/gasman/wagtail/tree/ababic-feature/readonlyfieldpanels - testing now.

**Comment by gasman** [comment]
Tested and all works well! Merged in b946b3d66d67e8cd7179c188078c6ed735c46b9c + parents.

File: wagtail/admin/panels/field_panel.py
Comment (ababic) [reply wagtail/admin/panels/field_panel.py:202]: I'm pretty sure I was just rewriting to better handle cases where `self.bound_field` is `None` (which is always the case for read-only panels)

**Comment by lb-** [comment]
@laymonage - this could possibly be used for an inspect style admin view for Snippets? You may already have plans underway, but just checking this on your radar.

**Comment by sowinski** [comment]
How can I make the slug field read only? I generate the slug problematically in my save method and don't want to make it editable?

```
    promote_panels = Page.promote_panels + [
        FieldPanel("slug", read_only=True)
    ]
```

**Pull request closed without merging**
